src/transformers/utils/hub.py

```diff
@@ -869,48 +870,110 @@ class PushToHubMixin:
+        use_auth_token: Optional[Union[bool, str]] = None,
         repo_url: Optional[str] = None,
-        use_temp_dir: bool = False,
-        commit_message: Optional[str] = None,
         organization: Optional[str] = None,
+    ):
+        """
+        Create the repo if needed, cleans up repo_id with deprecated kwards `repo_url` and `organization`, retrives the
+        token.
+        """
+        if repo_url is not None:
+            warnings.warn(
+                "The `repo_url` argument is deprecated and will be removed in v5 of Transformers. Use `repo_id` "
+                "instead."
+            )
+            repo_id = repo_url.replace(f"{HUGGINGFACE_CO_RESOLVE_ENDPOINT}/", "")
+        if organization is not None:
+            warnings.warn(
+                "The `organization` argument is deprecated and will be removed in v5 of Transformers. Set your "
+                "organization directly in the `repo_id` passed instead (`repo_id={organization}/{model_id}`)."
+            )
+            if not repo_id.startswith(organization):
+                if "/" in repo_id:
+                    repo_id = repo_id.split("/")[-1]
+                repo_id = f"{organization}/{repo_id}"
+
+        token = HfFolder.get_token() if use_auth_token is True else use_auth_token
+        url = create_repo(repo_id=repo_id, token=token, private=private, exist_ok=True)
+
+        # If the namespace is not there, add it or `upload_file` will complain
+        if "/" not in repo_id and url != f"{HUGGINGFACE_CO_RESOLVE_ENDPOINT}/{repo_id}":
+            repo_id = get_full_repo_name(repo_id, token=token)
+        return repo_id, token
+
+    def _get_files_timestamps(self, working_dir: Union[str, os.PathLike]):
+        """
+        Returns the list of files with their last modification timestamp.
+        """
+        return {f: os.path.getmtime(os.path.join(working_dir, f)) for f in os.listdir(working_dir)}
+
+    def _upload_modified_files(
+        self,
+        working_dir: Union[str, os.PathLike],
+        repo_id: str,
+        files_timestamps: Dict[str, float],
+        commit_message: Optional[str] = None,
+        token: Optional[str] = None,
+    ):
+        """
+        Uploads all modified files in `working_dir` to `repo_id`, based on `files_timestamps`.
+        """
+        modified_files = [
+            f
+            for f in os.listdir(working_dir)
+            if f not in files_timestamps or os.path.getmtime(os.path.join(working_dir, f)) > files_timestamps[f]
+        ]
+        for file in modified_files:
+            if commit_message is None:
+                commit_message = f"Upload {file}."
+            logger.info(f"Uploading {file} to {repo_id}")
+            url = upload_file(
+                path_or_fileobj=os.path.join(working_dir, file),
+                path_in_repo=file,
+                repo_id=repo_id,
+                token=token,
+                commit_message=commit_message,
+            )
```

Comment:
why not upload all files in one commit, using `create_commit` (or `upload_folder`)?

Reply:
Yes, this could be updated to leverage either of the two so that a single commit is done.

The `create_commit` docs are available [here](https://huggingface.co/docs/huggingface_hub/package_reference/hf_api#huggingface_hub.HfApi.create_commit).

The `upload_folder` docs are available [here](https://huggingface.co/docs/huggingface_hub/package_reference/hf_api#huggingface_hub.HfApi.upload_folder).

Reply:
If using `upload_folder`, you won't need to track modified files with their timestamps: if the files have the same name as others on the Hub and contain the same values, they won't get pushed (so if you have local checkpoints that have already been pushed it won't push them again, only verify if they're on the Hub).

Reply:
Added some comments to use `upload_folder` instead, it seems like it's an easy enough drop-in replacement

Reply:
`upload_folder` is not ideal here as we may be in a folder where the user has some other stuff (like training artifacts) they didn't explicitly asked us to push. But `create_commit` should work.

(I was asked to use `upload_file` looks like my instructions were not correct 😝  )

Reply:
seems like you have to take your "instructions" from the right persons=)

Reply:
🫣